.github/workflows/ci.yaml

```diff
@@ -34,3 +34,49 @@ jobs:
 
       - name: Run tests
         run: uv run pytest tests -m 'not integration'
+
+  # This runs integration tests of the OpenAI API (via the Ollama provider).
+  # This is an alternative to clouds, as testing those on PR will leak secrets.
+  ollama:
+    runs-on: ubuntu-latest
+
+    steps:
+      - uses: actions/checkout@v4
+
+      - name: Install UV
+        run: curl -LsSf https://astral.sh/uv/install.sh | sh
+
+      - name: Source Cargo Environment
+        run: source $HOME/.cargo/env
+
+      - name: Set up Python 3.12
+        run: uv python install 3.12
+
+      - name: Cache Ollama models
+        id: cache-ollama
+        uses: actions/cache@v4
+        with: # cache key is based on where OLLAMA_MODEL is defined.
+          path: ~/.ollama/models  # default directory for Ollama models
+          key: ollama-${{ hashFiles('./src/exchange/providers/ollama.py') }}
+
+      - name: Install Ollama
+        run: curl -fsSL https://ollama.com/install.sh | sh
+
+      - name: Start Ollama
+        run: |
+            # Run the background, in a way that survives to the next step
+            nohup ollama serve > ollama.log 2>&1 &
+
+            # Block using the ready endpoint mentioned in ollama/ollama#3341
+            time curl --retry 5 --retry-connrefused --retry-delay 0 -sf http://localhost:11434
+
+      # First time pull, and first time execution of a model is slow, so we do
```

Comment:
So, if running the tests via openai, they definitely won't pull (you get a 404.. there's no on-demand pulling). I'll clarify in another change this sentence since we won't cache anymore (that this is more about the pull vs first time pull)